DATA ANALYSIS PROJECT

In [65]:
#Importing Libraries 
import pandas as pd 
import matplotlib.pyplot as plt

In [66]:
df = pd.read_csv('sample.csv')

In [67]:
# Quick look into the dataframe: df
df.head()

recommendation_id           created_at   tasker_id  \
0  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009185352   
1  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1006892359   
2  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1012023956   
3  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009733517   
4  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1013579273   

   position  hourly_rate  num_completed_tasks  hired            category  
0         1           38                  151      0  Furniture Assembly  
1         2           40                  193      0  Furniture Assembly  
2         3           28                    0      0  Furniture Assembly  
3         4           43                  303      0  Furniture Assembly  
4         5           29                   39      0  Furniture Assembly

In [68]:
#Information on the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
recommendation_id      30000 non-null object
created_at             30000 non-null object
tasker_id              30000 non-null int64
position               30000 non-null int64
hourly_rate            30000 non-null int64
num_completed_tasks    30000 non-null int64
hired                  30000 non-null int64
category               30000 non-null object
dtypes: int64(5), object(3)
memory usage: 1.8+ MB


### Data Analysis

 1.) How many recommendation sets are in this data sample?

    To answer this question, let us find the unique recommendation ids in the data sample. 

In [69]:
df.recommendation_id.value_counts().count()

2100

There are 2100 recommendation sets in the given data sample.

 2.) Each recommendation set shows from 1 to 15 Taskers, what is:
  - average number of Taskers shown
  - median  number of Taskers shown

    To determine the average number of Taskers, let us calculate the mean using the mean function. 
    To determine the median number of taskers let us calculate the median using the median function.

In [70]:
df.recommendation_id.value_counts().mean()

14.285714285714286

In [71]:
df.recommendation_id.value_counts().median()

15.0

- The average number of Taskers shown in each recommendation set is 14.29
- The median number of Taksers shown in each recommendation set is 15.0

3.) How many total unique Taskers are there in this data sample?

    To answer this question, let us determine the unique Taskers using the value count and count function.

In [72]:
df.tasker_id.value_counts().count()

830

There are 830 unique Taskers in the given data sample.

4.) Which Tasker has been shown the most? Which Tasker has been shown the least?

In [73]:
shown_taskers = df.tasker_id.value_counts()

In [74]:
shown_taskers[shown_taskers == shown_taskers.max()]

1014508755    608
Name: tasker_id, dtype: int64

- The tasker_id: 1014508755 has been shown the most.

In [75]:
shown_taskers[shown_taskers == shown_taskers.min()]

1013573988    1
1009994950    1
1012386513    1
1007383273    1
1014926743    1
1009112003    1
1010009736    1
1014310300    1
1012678504    1
1012364558    1
1012166729    1
1010021990    1
1013573125    1
1011901532    1
1011949117    1
1010681878    1
1011968845    1
1013830691    1
1012348656    1
1009871933    1
1012289475    1
1006853970    1
1013934937    1
1006690425    1
1008469117    1
1009641175    1
1014547884    1
1010779242    1
1008604368    1
1009603880    1
             ..
1009772528    1
1014478773    1
1009754999    1
1008919567    1
1008368716    1
1013854788    1
1013656032    1
1013731883    1
1013362004    1
1010042971    1
1007246122    1
1011952623    1
1009547227    1
1008033678    1
1012805440    1
1007472083    1
1009618500    1
1011985968    1
1009702351    1
1007638825    1
1009712638    1
1007480912    1
1008870833    1
1010640007    1
1014593279    1
1014086818    1
1008474216    1
1012151299    1
1014439502    1
1011972750    1
Name: tasker_id, Length:

- 68 tasker_ids listed above have been shown least number of times. 

5.) Which Tasker has been hired the most? Which Tasker has been hired the least?

In [76]:
hired =df[df['hired'] ==1]['tasker_id'].value_counts()

In [77]:
hired[hired == hired.max()]

1012043028    59
Name: tasker_id, dtype: int64

- The tasker_id: 1012043028 has been hired the most.

In [78]:
hired[hired == hired.min()]

1009393887    1
1008966829    1
1013745838    1
1013434046    1
1007898815    1
1010008242    1
1007146669    1
1012719266    1
1013305557    1
1009729754    1
1014740602    1
1014832736    1
1013034579    1
1009848016    1
1008111352    1
1008903001    1
1009299585    1
1008037901    1
1014251534    1
1008782548    1
1010752503    1
1014478773    1
1014212647    1
1009733517    1
1012170634    1
1009638159    1
1008162664    1
1008762938    1
1010578972    1
1009348455    1
             ..
1009638573    1
1009703547    1
1013443125    1
1009192067    1
1009865854    1
1009693303    1
1008724560    1
1014629676    1
1010801075    1
1009751605    1
1009917428    1
1007955495    1
1014157578    1
1008887321    1
1013553854    1
1013852438    1
1013696131    1
1009606144    1
1008862713    1
1009966564    1
1011920226    1
1007702812    1
1008890855    1
1010438496    1
1008930827    1
1008008634    1
1008473496    1
1012229493    1
1013865107    1
1015020347    1
Name: tasker_id, Length:

- 79 takser_ids listed above have been hired the least.

6.) If we define the "Tasker conversion rate" as the number of times a
 Tasker has been hired, out of the number of times the Tasker has
 been shown, how many Taskers have a conversion rate of 100%

In [79]:
a = df.tasker_id.value_counts().to_frame()
a.columns = ['num_times_shown']

In [80]:
b = df.groupby('tasker_id')['hired'].sum().to_frame()
b.columns =['num_times_hired']

In [81]:
conversion_df=a.join(b)

In [82]:
conversion_df['rate'] = (conversion_df['num_times_hired']/ conversion_df['num_times_shown'])*100

In [83]:
conversion_df[conversion_df['rate'] == 100]

num_times_shown  num_times_hired   rate
1008861741                9                9  100.0
1008094420                2                2  100.0
1012369686                2                2  100.0
1014478773                1                1  100.0
1011985968                1                1  100.0
1007480912                1                1  100.0

In [84]:
conversion_df[conversion_df['rate'] == 100].count()

num_times_shown    6
num_times_hired    6
rate               6
dtype: int64

The number of taksers who have a 100% conversion rate is 6.

7.) Would it be possible for all Taskers to have a conversion rate of 100%
    Please explain your reasoning.

No, it is not possible for all Taksers to have a conversion rate of 100% because a "recommendation" is a group 
of Taskers (ranging from 1 to 15) from which the Client can choose ONE to book or rather only 1 tasker out of a set of recommendations can be hired. In this way, there will always be few tasksers who not be hired. 

For all Taskers to have a conversion rate of 100%, then only one taskers must be recommended for a task and the takser must be hired. 



8.) For each category, what is the average position of the Tasker who is hired?

In [85]:
df[df['hired']==1].groupby('category')['position'].mean()

category
Furniture Assembly    3.611888
Mounting              4.596085
Moving Help           4.145359
Name: position, dtype: float64

    Since the position cannot be a decimal value. Lets take the floor value.

In [86]:
import numpy as np

In [87]:
np.floor(df[df['hired']==1].groupby('category')['position'].mean())

category
Furniture Assembly    3.0
Mounting              4.0
Moving Help           4.0
Name: position, dtype: float64

- The average position of the Tasker who is hired for Furniture Assembly is 3
- The average position of the Tasker who is hired for Mounting is 4
- The average position of the Tasker who is hired for Moving help is 4

9.) For each category, what is the average hourly rate and average number of completed tasks for the Taskers who are hired?

In [88]:
market_data=df[df['hired']==1].groupby('category')['hourly_rate', 'num_completed_tasks'].mean()
market_data

hourly_rate  num_completed_tasks
category                                            
Furniture Assembly    38.701049           249.020979
Mounting              50.154804           284.096085
Moving Help           63.012259           273.882662

    Since the number of completed tasks cannot be a decimal value. Lets take the floor value.

In [89]:
market_data['num_completed_tasks'] = np.floor(market_data['num_completed_tasks'])
market_data

hourly_rate  num_completed_tasks
category                                            
Furniture Assembly    38.701049                249.0
Mounting              50.154804                284.0
Moving Help           63.012259                273.0

The above dataframe show the average hourly rate and average number of completed tasks for the Taskers who were hired in each catgeory.

10.) Based on the previous, how would you approach the question of: How can we use market data to suggest hourly rates to Taskers that would maximize their opportunity to be hired? Please describe in detail, with code and formulas that support your model.


    Idea/Concept:  I have based my solution on the idea that taskers who were hired exhibit a normal behavior 
    in each category. All tasksers who are a part of this distribution will have an average hourly rate 
    as described in the market data Data Frame.To maximize the opportunity to be hired 
    I will suggest hourly rates for taskers that are a part of the normal distribution in each category.

    Approach: I have used the z-score method to detect if the z-score value of the hourly rate of a tasker, who is not hired, is greater than 3 or less than -3. If this condition is satisfied, then that tasker would be considered as an outlier to the distribution. So, this particular takser's hourly rate would be reassigned to the average hourly rate of takser who is hired for a particular catgeory.

    Formula to find the z-score: z = (X - mean)/(Standard deviation) 
    - where X is the hourly rate of the tasker
    - mean is the mean of the taskers hourly rate of each category
    - Standard deviation is the standard deviation of the hourly rate of each category. 
    
    Condition for an outlier: abs(z) > 3 

    Lets calculate the standard deviation for takser's hourly rate, which will be used in z-score calculation. 

In [90]:
df[df['hired']==1].groupby('category')['hourly_rate'].std()

category
Furniture Assembly    13.514984
Mounting              10.043536
Moving Help           29.959649
Name: hourly_rate, dtype: float64

    Let us now implement the approach discussed above.

In [91]:
def hourly_rate_for_optimal_hiring(hired, category, hourly_rate):
    z= 0 # here we define the Value of z score, set to 0 intially.   
    if hired == 0: # Here We will consider tasker who have not been hired. 
        # Next, For taskers not hired, we will calculate z score of hourly rate for each category. 
        if category == 'Furniture Assembly':
            z = (hourly_rate - 38.701049) / 13.514984
            # if z-score of taskers hourly rate is greater than 3 or less than -3 then, 
            # replace the score by the mean of the category hourly rate. 
            if abs(z) > 3: 
                hourly_rate = 38.701049
        elif category == 'Mounting':
            z = (hourly_rate - 50.154804) / 10.043536
            if abs(z) > 3: 
                hourly_rate = 50.154804
        else:
            z = (hourly_rate - 63.012259) / 29.959649
            if abs(z) > 3: 
                hourly_rate = 63.012259
# If the tasker has been hired then that was his 
#optimal hourly rate, so we will keep it the same. 
    else:
        hourly_rate = hourly_rate
                
    return hourly_rate

In [92]:
hourly_rate_optimal_hiring = []

In [93]:
# Lets, apply the function to our data. and store the result in a list.  
for index, row  in df.iterrows():
    hourly_rate_optimal_hiring.append(hourly_rate_for_optimal_hiring(row['hired'], row['category'], row['hourly_rate']))
    

In [94]:
# Lets, create a new column in our data frame 'df' which suggests the optimal hourly rate. 
df['optimum_hourly_rate'] = hourly_rate_optimal_hiring

In [95]:
df.head()

recommendation_id           created_at   tasker_id  \
0  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009185352   
1  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1006892359   
2  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1012023956   
3  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009733517   
4  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1013579273   

   position  hourly_rate  num_completed_tasks  hired            category  \
0         1           38                  151      0  Furniture Assembly   
1         2           40                  193      0  Furniture Assembly   
2         3           28                    0      0  Furniture Assembly   
3         4           43                  303      0  Furniture Assembly   
4         5           29                   39      0  Furniture Assembly   

   optimum_hourly_rate  
0                 38.0  
1                 40.0  
2                 28.0  
3                 43.0  
4                 29.0